state/cluster/badger/mutator_test.go

```diff
@@ -396,6 +385,72 @@ func (suite *MutatorSuite) TestExtend_WithReferenceBlockFromClusterChain() {
 	suite.Assert().Error(err)
 }
 
+// TestExtend_WithReferenceBlockFromDifferentEpoch tests extending the cluster state
+// using a reference block in a different epoch than the cluster's epoch.
+func (suite *MutatorSuite) TestExtend_WithReferenceBlockFromDifferentEpoch() {
+	// build and complete the current epoch, then use a reference block from next epoch
+	eb := unittest.NewEpochBuilder(suite.T(), suite.protoState)
+	eb.BuildEpoch().CompleteEpoch()
+	heights, ok := eb.EpochHeights(1)
+	require.True(suite.T(), ok)
+	nextEpochHeader, err := suite.protoState.AtHeight(heights.FinalHeight() + 1).Head()
+	require.NoError(suite.T(), err)
+
+	block := suite.Block()
+	block.SetPayload(model.EmptyPayload(nextEpochHeader.ID()))
+	err = suite.state.Extend(&block)
+	suite.Assert().Error(err)
+	suite.Assert().True(state.IsInvalidExtensionError(err))
+}
+
+// TestExtend_WithUnfinalizedReferenceBlock tests that extending the cluster state
+// with a reference block which is un-finalized and above the finalized boundary
+// should be considered an unverifiable extension. It's possible that this reference
+// block has been finalized, we just haven't processed it yet.
+func (suite *MutatorSuite) TestExtend_WithUnfinalizedReferenceBlock() {
+	unfinalized := unittest.BlockWithParentFixture(suite.protoGenesis)
+	unfinalized.Payload.Guarantees = nil
+	unfinalized.SetPayload(*unfinalized.Payload)
+	err := suite.protoState.ExtendCertified(context.Background(), unfinalized, unittest.CertifyBlock(unfinalized.Header))
+	suite.Require().NoError(err)
+
+	block := suite.Block()
+	block.SetPayload(model.EmptyPayload(unfinalized.ID()))
+	err = suite.state.Extend(&block)
+	suite.Assert().Error(err)
+	suite.Assert().True(state.IsUnverifiableExtensionError(err))
+}
+
+// TestExtend_WithOrphanedReferenceBlock tests that extending the cluster state
+// with a reference block is un-finalized and below the finalized boundary
+// (i.e. orphaned) should be considered an invalid extension. This reference block
+// can never be finalized, therefore the proposer knowingly generated an invalid
+// cluster block proposal.
+func (suite *MutatorSuite) TestExtend_WithOrphanedReferenceBlock() {
+	// create a block extending genesis which is not finalized
+	orphaned := unittest.BlockWithParentFixture(suite.protoGenesis)
+	orphaned.Payload.Guarantees = nil
```

Comment:
`BlockWithParentFixture` creates an empty payload, am I missing something?

Reply:
Yes, this was unecessary -- removed